### PR TITLE
add nacos support

### PR DESCRIPTION
**Pull request by afghanistanyn**
### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible?

### CODE REVIEW

File: doc/zh-cn/discovery-nacos.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+```json
+{
+	"apisix.gateway.registration": "false"
+}
+```
+---
+
+#### 为什么只支持指定的namespace和group
+
+1. 获取nacos服务列表以及服务详情必须指定namespace或group,也无获取命名空间相关api, 但可以通过配置指定解决
```
Comment (membphis) [inline doc/zh-cn/discovery-nacos.md:70]: need a space between Chinese and English

File: apisix/discovery/nacos.lua [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        headers['Content-Type'] = 'application/json'
+    end
+
+    local httpc = http.new()
+    local timeout = local_conf.nacos.timeout
+    local connect_timeout = timeout and timeout.connect or 2000
+    local send_timeout = timeout and timeout.send or 2000
+    local read_timeout = timeout and timeout.read or 5000
+    log.info("connect_timeout:", connect_timeout, ", send_timeout:", send_timeout,
+            ", read_timeout:", read_timeout, ".")
```
Comment (membphis) [inline apisix/discovery/nacos.lua:126]: bad indentation

File: apisix/discovery/nacos.lua [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        query = query,
+        body = body,
+        ssl_verify = false,
+    })
+end
+
+
+local function fetch_instance(service)
+    if not service then
+	return
```
Comment (membphis) [inline apisix/discovery/nacos.lua:141]: bad indentation

File: apisix/discovery/nacos.lua [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        -- add serviceName and weight
+        -- weigth is required , when get node for service by apisix
+        local innerInstanceInfo = core.table.new(0, #host + 2)
+        innerInstanceInfo["serviceName"] = host.serviceName
+        innerInstanceInfo["weight"] = host.weight or  host.metadata["weight"]  or default_weight
+        innerInstanceInfo["host"]= host.ip
+        innerInstanceInfo["port"] = host.port
+        innerInstanceInfo["metadata"] = host.metadata
+        if innerInstanceInfo["metadata"]["apisix.gateway.registration"] ~= "false" then
+                core.table.insert(instanceInfo, innerInstanceInfo)
```
Comment (membphis) [inline apisix/discovery/nacos.lua:192]: bad indentation

File: apisix/discovery/nacos.lua [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+        timeout = {
+            type = "object",
+            properties = {
+                connect = {type = "integer", minimum = 1, default = 2000},
+                send = {type = "integer", minimum = 1, default = 2000},
+                read = {type = "integer", minimum = 1, default = 5000},
+            }
+        },
+    },
+    required = {"host"}
```
Comment (membphis) [inline apisix/discovery/nacos.lua:60]: add `additionalProperties = true`

**Comment by membphis** [comment]
@afghanistanyn please add test cases

File: apisix/discovery/nacos.lua [hunk 6]
```diff
@@ -165,7 +165,7 @@ local function fetch_instance(service)
     local json_str = res.body
     local data, err = core.json.decode(json_str)
     if not data then
-        log.error("get instance from nacos for " .. service  .. ", invalid response body: ", json_str, " err: ", err)
+        log.error("get instance for "..service..",invalid response body: ",json_str, " err: ", err)
```
Comment (membphis) [inline apisix/discovery/nacos.lua:168]: that is not a good error log style.
we need to avoid to concat string object, bad performance

Comment (afghanistanyn) [reply apisix/discovery/nacos.lua:168]: help 
could you help correct it and add some test cases， it’s too hard for me

**Comment by moonming** [comment]
> ### What this PR does / why we need it:
> ### Pre-submission checklist:
> 
>     * [ ]  Did you explain what problem does this PR solve? Or what new features have been added?
> 
>     * [ ]  Have you added corresponding test cases?
> 
>     * [ ]  Have you modified the corresponding document?
> 
>     * [ ]  Is this PR backward compatible?

@afghanistanyn please take a look at this checklist first.

**Comment by moonming** [comment]
@afghanistanyn Compared with the dns mode of nacos, what are the advantages of this?

File: apisix/discovery/nacos.lua [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+    end
+    if local_conf.nacos.prefix then
+        url = url .. local_conf.nacos.prefix
+    end
+    if string_sub(url, #url) ~= "/" then
+        url = url .. "/"
+    end
+
+    return url, basic_auth
+end
```
Comment (moonming) [inline apisix/discovery/nacos.lua:96]: This function is basically the same as eureka, please make it a public function.

**Comment by HuangDayu** [comment]
如何使用nacos做服务发现呢？而不是手动设置服务。

File: doc/zh-cn/discovery-nacos.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  prefix: "/nacos/"         
+  namespace: public         //默认命名空间
+  group: DEFAULT_GROUP      //默认分组
+  fetch_interval: 30
+  weight: 100
+  timeout:
+    connect: 2000
+    send: 2000
+    read: 5000
+
```
Comment (nic-chen) [inline doc/zh-cn/discovery-nacos.md:53]: Could you give an example of configuring a specific service, please?

Including `route` and `upstream` config.

**Pull request closed without merging**